ipykernel/kernelbase.py

```diff
@@ -356,28 +361,91 @@ async def advance_eventloop():
     def _message_counter_default(self):
         return itertools.count()
 
-    async def shell_main(self):
+    async def shell_channel_thread_main(self):
+        """Main loop for shell channel thread.
+
+        Listen for incoming messages on kernel shell_socket.  For each message
+        received, extract the subshell_id from the message header and forward the
+        message to the correct subshell via ZMQ inproc pair socket.
+        """
+        assert self.shell_socket is not None
+        assert self.session is not None
+        assert self.shell_channel_thread is not None
+        assert threading.current_thread() == self.shell_channel_thread
+
+        try:
+            while True:
+                msg = await self.shell_socket.recv_multipart()
+
+                # Deserialize whole message just to get subshell_id.
+                # Keep original message to send to subshell_id unmodified.
+                # Ideally only want to deserialize message once.
+                copy = not isinstance(msg[0], zmq.Message)
+                _, msg2 = self.session.feed_identities(msg, copy=copy)
+                try:
+                    msg3 = self.session.deserialize(msg2, content=False, copy=copy)
+                    subshell_id = msg3["header"].get("subshell_id")
+
+                    # Find inproc pair socket to use to send message to correct subshell.
+                    socket = self.shell_channel_thread.manager.get_shell_channel_socket(subshell_id)
+                    assert socket is not None
+                    socket.send_multipart(msg, copy=False)
+                except Exception:
```

Comment:
Do you have any idea if we can narrow this exception, or can it be anything, or maybe too many possibilities ?

Reply:
I don't know. This is the same usage as in the existing code at
https://github.com/ipython/ipykernel/blob/0fa543936d3f9ee69f6b9f8bd5a316692fde9aaa/ipykernel/debugger.py#L442-L445
so I am assuming that a failed `deserialize` could raise any `Exception`, and I've put the forwarding of the message in the same `try` block as if it is not a valid message it should not be forwarded.

Reply:
Sounds good. Let's leave `Exception` then.

Reply:
Left as it is.